zerver/lib/test_helpers.py

```diff
@@ -598,3 +602,27 @@ def zulip_reaction_info() -> Dict[str, str]:
         emoji_code='zulip',
         reaction_type='zulip_extra_emoji',
     )
+
+def create_email_change_confirmation_key(user_profile: UserProfile, new_email: str,
+                                         email_address_visibility: int=Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE,
+                                         date_sent: Optional[datetime]=None) -> str:
+    do_set_realm_property(
```

Comment:
A function named like this shouldn't sneakily change realm properties. If tests need email visibility changed (why btw?) they should do it explicitly. Also, seems like `create_email_change_confirmation_key` is the kind of code that should be shared between tests and the production code that does the same thing, rather than having two places in the code that aim to do the same thing.

Reply:
> why btw?

There's a test in the previous tests called `test_change_delivery_email_end_to_end_with_admins_visibility`. It is for that and the very few similar tests. Changed it to have it in the actual test's code.

I think we cannot really use the production code. We create the URL and mail it to the user in a single function  `do_start_email_change_process`. So we cannot directly get the URL/key from production code. and another issue would be us setting the required date to test expired key.

Reply:
Ahh, we default to `EMAIL_ADDRESS_VISIBILITY_ADMINS`, now removing that is causing the coverage issue 😬

Reply:
Changed `test_social_auth_email_change_with_admins_visibility` to `test_social_auth_email_change_delivery_email_public` and moved the assertion that checks if `email` isn't changed and remained as user id to the `test_social_auth_email_change_success` test.
[Here](https://pastebin.com/kUQrz9Uz)'s the diff for that fix.

Reply:
> I think we cannot really use the production code. We create the URL and mail it to the user in a single function do_start_email_change_process. So we cannot directly get the URL/key from production code. and another issue would be us setting the required date to test expired key.

I'd say we should extract the common "create the link" logic into a function, so that it can used in both `start_email_change_process` and in tests, it can take `date_sent` as an argument.

Reply:
"create the link" is already a different function(`create_confirmation_link`) which generates the key and returns URL with key in it. 
For quick reference here's the production code
```python
def do_start_email_change_process(user_profile: UserProfile, new_email: str) -> None:
    old_email = user_profile.delivery_email
    obj = EmailChangeStatus.objects.create(new_email=new_email, old_email=old_email,
                                           user_profile=user_profile, realm=user_profile.realm)

    activation_url = create_confirmation_link(obj, Confirmation.EMAIL_CHANGE)
    from zerver.context_processors import common_context
    context = common_context(user_profile)
    context.update({
        'old_email': old_email,
        'new_email': new_email,
        'activate_url': activation_url,
    })
    language = user_profile.default_language
    send_email('zerver/emails/confirm_new_email', to_emails=[new_email],
               from_name=FromAddress.security_email_from_name(language=language),
               from_address=FromAddress.tokenized_no_reply_address(),
               language=language, context=context,
               realm=user_profile.realm)
```
I'm unable to see a way to break that so we could use it for test code which would let us set our own `date_sent`. :disappointed:

Reply:
Call `create_confirmation_link` in the test helper, and change `date_sent` field on the created object to the intended one? :thinking:

Reply:
Hmm, that way we'd just save this statement 
```
    Confirmation.objects.create(content_object=obj,
                                date_sent=date_sent,
                                confirmation_key=key,
                                type=Confirmation.EMAIL_CHANGE)
```
in(just for reference)
```python
def create_email_change_confirmation_key(user_profile: UserProfile, new_email: str,
                                         date_sent: Optional[datetime.datetime]=None) -> str:
    old_email = user_profile.delivery_email
    new_email = new_email
    obj = EmailChangeStatus.objects.create(new_email=new_email,
                                           old_email=old_email,
                                           user_profile=user_profile,
                                           realm=user_profile.realm)
    key = generate_key()
    if date_sent is None:
        date_sent = timezone_now()
    Confirmation.objects.create(content_object=obj,
                                date_sent=date_sent,
                                confirmation_key=key,
                                type=Confirmation.EMAIL_CHANGE)
    return key
```
and will have to add a line to retrieve `action_key` from the link to send it in as GET/POST's for auth flows along with adding the code for changing `date_sent`.

Reply:
```
    obj = EmailChangeStatus.objects.create(new_email=new_email, old_email=old_email,
                                           user_profile=user_profile, realm=user_profile.realm)

    activation_url = create_confirmation_link(obj, Confirmation.EMAIL_CHANGE)
```

this can be extracted from the prduction code into a `create_email_change_confirmation_link`  helper, and then the test will call that create_email_change_confirmation_link helper and alter the date_sent and grab the key :thinking:  It might not save many, or any, actual lines of code but the point is to encapsulate the details of the procedure of creating this object, so I think it's worth it nonetheless

Reply:
> It might not save many, or any, actual lines of code but the point is to encapsulate the details of the procedure of creating this object, so I think it's worth it nonetheless


Now, it very clear to me why we wanted to use the production code. Thanks, made the changes as a prep commit.